Use delimiter, null and quote from sql command

Greenscreen23 opened the pull request:
This PR implements the DELIMITER, NULL and QUOTE sql options for importing CSV files in hyrise. The goal is that a query like this is executed correctly
```
COPY "{}" FROM '{}' WITH (FORMAT CSV, DELIMITER '|', NULL '', QUOTE '"');
```

Fixes #2747

Greenscreen23 committed:
Update sql-parser

Greenscreen23 committed:
My code is working, but I broke other code

Greenscreen23 commented:
@dey4ss I upgraded the SQL parser and now some tests regarding NULLS FIRST/LAST fail. This is because my upgrade also included one of your commits implementing this feature in the parser. The hyrise code has the following comment:
```
    // (iii) As of now, Hyrise supports only NULLS FIRST. Make sure that we notice when the parser supports explicitly
    //       specifying NULLS FIRST/LAST. If it does, we fail here so we know we have to adapt the SQLTranslator code.
```
Does that mean that this PR is blocked by this other feature? Is there already a branch for this? Should I try to implement this in this PR?

Greenscreen23 committed:
Throw an InvalidInputException when users specify NULLS LAST

Bouncner commented:
Can you add a proper description to the PR, please?

Bouncner reviewed (commented):
Still have to read the actual implementation. But I think cleaning up the `hsql` stuff should be done first.

Greenscreen23 committed:
First batch

Greenscreen23 committed:
Move to sql_to_lqp_helper

Bouncner commented on src/test/lib/sql/sql_translator_test.cpp:
```diff
@@ -1341,14 +1345,18 @@ TEST_F(SQLTranslatorTest, OrderByNullOrdering) {
+        "SELECT * FROM int_float ORDER BY a ASC NULLS LAST",
+        "SELECT * FROM int_float ORDER BY a DESC NULLS LAST",
+    };
+    for (const auto& query : invalid_queries) {
```
Could you do something like:
```
for (const auto& query : std::initialize_list<std::string>{"SELECT * FROM int_float ORDER BY a NULLS LAST",
                                                           "SELECT * FROM int_float ORDER BY a ASC NULLS LAST", ...
         ) {
...
```
But seeing it written out, I am not sure if that's any better. :D

Bouncner reviewed (commented):
.

Greenscreen23 committed:
simplify test

Bouncner commented:
Can you please add a test using a header-less version of `resources/test_data/tbl/tpch/sf-0.002/lineitem.tbl` and check if we can properly load it via SQL CSV import?

Table creation:
```
CREATE TABLE lineitem ( l_orderkey INTEGER not null, l_partkey INTEGER not null, l_suppkey INTEGER not null, l_linenumber INTEGER not null, l_quantity DECIMAL(12,2) not null, l_extendedprice DECIMAL(12,2) not null, l_discount DECIMAL(12,2) not null, l_tax  DECIMAL(12,2) not null, l_returnflag CHAR(1) not null, l_linestatus CHAR(1) not null, l_shipdate DATE not null, l_commitdate DATE not null, l_receiptdate DATE not null, l_shipinstruct CHAR(25) not null, l_shipmode CHAR(10) not null, l_comment VARCHAR(44) not null );
```

Greenscreen23 committed:
Fix separator delimiter confusion

Greenscreen23 committed:
Add test whether TPCH lineitem is correctly read, with additional fixes:
- lineitem.tbl gets a type row
- insert.cpp does not append an empty chunk when it should not append any rows (as this becomes immutable when copy from is called, which is not possible for empty chunks)

Bouncner commented on src/lib/operators/insert.cpp:
```diff
@@ -126,7 +126,7 @@ std::shared_ptr<const Table> Insert::_on_execute(std::shared_ptr<TransactionCont
     auto remaining_rows = left_input_table()->row_count();
 
-    if (_target_table->chunk_count() == 0) {
+    if (_target_table->chunk_count() == 0 && remaining_rows > 0) {
```
@dey4ss: Pinging you here.

Greenscreen23 replied:
Well, the pipeline certainly says this is not an issue.

Bouncner commented on src/test/lib/import_export/csv/csv_options_test.cpp:
```diff
@@ -84,4 +84,24 @@ TEST_F(CsvOptionsTest, ParseNull) {
+      "COPY lineitem FROM 'resources/test_data/csv/tpch/lineitem.csv' WITH (FORMAT CSV, DELIMITER '|')";
+  const auto [status, tables] = SQLPipelineBuilder(sql_statement).create_pipeline().get_result_tables();
+  EXPECT_EQ(status, SQLPipelineStatus::Success);
+  EXPECT_TABLE_EQ(Hyrise::get().storage_manager.get_table("lineitem"), expected_table, OrderSensitivity::Yes, TypeCmpMode::Strict, FloatComparisonMode::AbsoluteDifference);
```
Ah, now I see why you added the changes above.

Can you please use 0.001 here?

Greenscreen23 committed:
Small fixes

Bouncner commented on src/lib/sql/sql_translator.cpp:
```diff
@@ -304,6 +304,31 @@ FrameBound translate_frame_bound(const hsql::FrameBound& hsql_frame_bound) {
+  }
+  auto csv_parse_config = ParseConfig{};
+  if (csv_options->delimiter) {
+    AssertInput(strnlen(csv_options->delimiter, 2) == 1, "CSV delimiter should be exactly one char.");
```
Since char is a data type: "one character"?

Bouncner commented on src/lib/sql/sql_translator.cpp:
```diff
@@ -304,6 +304,31 @@ FrameBound translate_frame_bound(const hsql::FrameBound& hsql_frame_bound) {
+    csv_parse_config.quote = csv_options->quote[0];
+  }
+  if (csv_options->null) {
+    auto len = strnlen(csv_options->null, 100);
```
The 100 looks very arbitrary. What's your idea here?

Bouncner replied:
Is it tested?

Bouncner commented on src/test/lib/import_export/csv/csv_options_test.cpp:
```diff
@@ -0,0 +1,109 @@
+namespace hyrise {
+class CsvOptionsTest : public BaseTest {};
+
+TEST_F(CsvOptionsTest, ParseDelimier) {
```
ParseDelimier

Greenscreen23 committed:
Add test to check whether querying a table without chunks works

Greenscreen23 committed:
Assert that insert operators are only called with rows to use

Greenscreen23 committed:
Add test for TPCH region

Greenscreen23 commented:
> Sorry for annoying you once again, but I just realized that our "TPC-H test data" is not the actual TPC-H test data.
> 
> Can you please have a quick look what we need to change in order to load `region.tbl`:
> 
> ```
> 0|AFRICA|lar deposits. blithely final packages cajole. regular waters are final requests. regular accounts are according to |
> 1|AMERICA|hs use ironic, even requests. s|
> 2|ASIA|ges. thinly even pinto beans ca|
> 3|EUROPE|ly final courts cajole furiously final excuse|
> 4|MIDDLE EAST|uickly special accounts cajole carefully blithely close requests. carefully final asymptotes haggle furiousl|
> ```
> 
> `CREATE TABLE region ( r_regionkey INTEGER not null, r_name CHAR(25) not null, r_comment VARCHAR(152) );`
> 
> `COPY "region" FROM 'tpch-dbgen/sf-10.0/region.tbl' WITH (FORMAT CSV);`

There were two things that needed to be adapted:
- The r_comment row was nullable in the csv/sql import and not null in the tbl import. I have changed the sql, so it is always not null.
- Our csv parser does not like it when csv rows terminate with a delimiter. I stripped the last `|` from all rows.
Then it imports correctly :)

Greenscreen23 committed:
Merge branch 'master' of github.com:hyrise/hyrise into lukas/copy-from-sql

Greenscreen23 commented:
> > There were two things that needed to be adapted:
> > 
> > * The r_comment row was nullable in the csv/sql import and not null in the tbl import. I have changed the sql, so it is always not null.
> 
> Not sure what you mean here. The `tbl` does not contain any null information.
> 
> > * Our csv parser does not like it when csv rows terminate with a delimiter. I stripped the last `|` from all rows.
> >   Then it imports correctly :)
> 
> Yeah, I know that stripping the last delimiter works. That's why the TPC-H files are stored like that in our repository. However, the files are actually generated like that by the official data generator for TPC-H.
> 
> The above region.tbl is the exact content as generated.

I am unsure of what you want me to do. I have already created a separate test for this file, where it (with the training `|` stripped) is correctly loaded. Should I adapt the parser to allow for trailing delimiters at the end of a line?

Bouncner commented:
> Should I adapt the parser to allow for trailing delimiters at the end of a line?

yes, please

dey4ss commented:
>> -  Our csv parser does not like it when csv rows terminate with a delimiter. I stripped the last | from all rows.
        Then it imports correctly :)
>
>Yeah, I know that stripping the last delimiter works. That's why the TPC-H files are stored like that in our repository.
However, the files are actually generated like that by the official data generator for TPC-H.

Don't we just screw up correct CSV parsing then (https://datatracker.ietf.org/doc/html/rfc4180)? Separator at the end should mean that there is an extra `NULL` value at the end. I don't really think we should default to support some TPC-H specific encoding that supposedly is not even a CSV file. The downside of just supporting anything not RFC-complying is usually that weird things happen with dirty CSVs and we do not even notice it.

Bouncner commented:
It's a perfectly valid CSV file (well, TBL file with | delimiters). It's "just the interpretation" when imported into tables. I wouldn't say the files break any specifications.
It's basically THE standard format for TPC-H, DS and thus also SSB. We test data loading right now into other databases and so far (haven't tested them all) none of the other systems has issues with loading those tbl files.

There is no standard saying that importing a 5 column CSV into a three column table wouldn't be possible. Not that I like the TPC-H format, quite the opposite. But it's argue to say all other databases just work fine for **the** standard data generator files, but we don't.

Also having TPC-H test files for tests that aren't real TPC-H files but modified ones is ... strange.

Greenscreen23 commented:
How about a warning when this happens? Alternatively, we could tell the parser when the file ends with `.tbl` and only then accept trailing delimiters?

Bouncner commented:
Let’s skip it in this PR. Don’t want to block.

Sorry for the confusion. :(

dey4ss commented on src/lib/logical_query_plan/export_node.hpp:
```diff
@@ -13,12 +14,14 @@ namespace hyrise {
  public:
-  ExportNode(const std::string& init_file_name, const FileType init_file_type);
+  ExportNode(const std::string& init_file_name, const FileType init_file_type,
+             std::optional<ParseConfig> init_csv_parse_config = std::nullopt);
```
Now that `ParseConfig` is all over the place, we should rename it to make it more obvious it's about CSVs (or pack the CSV meta stuff in an own namespace).

Greenscreen23 replied:
I renamed it to `CsvParseConfig`. I don't think there is enough content to justify an entire namespace.

dey4ss commented on src/lib/operators/maintenance/create_table.cpp:
```diff
@@ -85,10 +85,13 @@ std::shared_ptr<const Table> CreateTable::_on_execute(std::shared_ptr<Transactio
-    _insert->set_transaction_context(context);
-    _insert->execute();
+    // Insert table data only if there is data. We would generate an empty chunk if we would call this operator without
+    // this check, which would crash when data is appended to the table via copy from.
```
```suggestion
    // this check, which would crash when data is appended to the table via a `COPY FROM` statement.
```

dey4ss commented on src/lib/operators/maintenance/create_table.cpp:
```diff
@@ -85,10 +85,13 @@ std::shared_ptr<const Table> CreateTable::_on_execute(std::shared_ptr<Transactio
+    // Insert table data only if there is data. We would generate an empty chunk if we would call this operator without
+    // this check, which would crash when data is appended to the table via copy from.
+    if (_left_input->get_output()->row_count() > 0) {
+      _insert = std::make_shared<Insert>(table_name, _left_input);
```
Is there a need to have the `Insert` operator as a member?

Greenscreen23 replied:
I don't think so, it compiles without it being a member. Are there any other cases of operators except these three (`_insert` in `create_table.cpp`, `_insert` and `_delete` in `update.cpp`)? I did not find anything by searching for `_insert;` and `_delete;`

dey4ss commented on src/lib/operators/update.cpp:
```diff
@@ -49,11 +49,14 @@ std::shared_ptr<const Table> Update::_on_execute(std::shared_ptr<TransactionCont
+    _insert = std::make_shared<Insert>(_table_to_update_name, _right_input);
+    _insert->set_transaction_context(context);
+    _insert->execute();
+    // Insert cannot fail in the MVCC sense, no check necessary
```
```suggestion
    // Insert cannot fail in the MVCC sense, no check necessary.
```

dey4ss commented on src/lib/operators/update.cpp:
```diff
@@ -49,11 +49,14 @@ std::shared_ptr<const Table> Update::_on_execute(std::shared_ptr<TransactionCont
+  // 2. Insert new data with the Insert operator only if there is data. We would generate an empty chunk if we would
+  // call this operator without this check, which would crash when data is appended to the table via copy from.
+  if (_right_input->get_output()->row_count() > 0) {
+    _insert = std::make_shared<Insert>(_table_to_update_name, _right_input);
```
See my comment on `CreateTable`. Might also apply to the `Delete`operator here.

dey4ss commented on src/lib/sql/sql_translator.cpp:
```diff
@@ -304,6 +304,31 @@ FrameBound translate_frame_bound(const hsql::FrameBound& hsql_frame_bound) {
+    auto len = strnlen(csv_options->null, 100);
+    AssertInput(csv_options->null[len] == '\0', "CSV null string cannot be larger than 100 characters.");
+    csv_parse_config.null_string = std::string(csv_options->null, len);
+    boost::to_lower(csv_parse_config.null_string);
```
Is there a reason to assert the length and not just use the `std::string` constructor without length (7 here: https://en.cppreference.com/w/cpp/string/basic_string/basic_string.html)?

Also, I am not sure about the case conversion here.

dey4ss reviewed (commented):
Haven't looked at the tests, yet. I have some questions, though.

Greenscreen23 committed:
Daniel's suggestions

Bouncner commented on resources/test_data/csv/null_as_string_with_bad_null.csv:
```diff
@@ -1,3 +1,3 @@
 "Hallo",458,12345
-NULL,123,456
```
Do we test the discussed case-sensitive matching somewhere?

Bouncner commented on src/lib/operators/maintenance/create_table.cpp:
```diff
@@ -85,10 +85,13 @@ std::shared_ptr<const Table> CreateTable::_on_execute(std::shared_ptr<Transactio
-    _insert->set_transaction_context(context);
-    _insert->execute();
+    // Insert table data only if there is data. We would generate an empty chunk if we would call this operator without
+    // this check, which would crash when data is appended to the table via a `COPY FROM` statement.
```
The comment does not really help  me.

Greenscreen23 replied:
I've reworded it, does that help?

Bouncner commented on src/lib/operators/insert.cpp:
```diff
@@ -125,6 +125,7 @@ std::shared_ptr<const Table> Insert::_on_execute(std::shared_ptr<TransactionCont
     const auto append_lock = _target_table->acquire_append_mutex();
 
     auto remaining_rows = left_input_table()->row_count();
+    Assert(remaining_rows > 0, "Insert operator should only be called if there are rows to insert");
```
```suggestion
    Assert(remaining_rows > 0, "Insert operator should only be called if there are rows to insert.");
```

Bouncner commented on src/lib/operators/update.cpp:
```diff
@@ -39,21 +39,24 @@ std::shared_ptr<const Table> Update::_on_execute(std::shared_ptr<TransactionCont
-  _insert->execute();
-  // Insert cannot fail in the MVCC sense, no check necessary
+  // 2. Insert new data with the Insert operator only if there is data. We would generate an empty chunk if we would
+  // call this operator without this check, which would crash when data is appended to the table via copy from.
```
See above.

Bouncner reviewed (commented):
Need to check the tests.

Greenscreen23 committed:
Martin's things

Greenscreen23 committed:
Test that uppercase null strings are not null

Bouncner commented on src/test/lib/operators/insert_test.cpp:
```diff
@@ -513,4 +513,15 @@ TEST_F(OperatorsInsertTest, MarkMultipleChunksImmutable) {
   }
 }
 
+TEST_F(OperatorsInsertTest, QueryTableWithoutChunks) {
```
What is the purpose of this "insert" test?

Bouncner commented on src/test/lib/import_export/csv/csv_options_test.cpp:
```diff
@@ -0,0 +1,128 @@
+      "quote_table", std::make_shared<Table>(table_column_definitions, TableType::Data, std::nullopt, UseMvcc::Yes));
+
+  const auto sql_statement =
+      std::string{"COPY quote_table FROM 'resources/test_data/csv/int_string_quoted.csv' WITH (FORMAT CSV, QUOTE '`')"};
```
Comparing different tables?

Bouncner committed:
Merge, merge, merge

Bouncner committed:
Lint fix
